## Exercício 6




### Conectando com a base

In [77]:
import matplotlib.pyplot as plt
import pandas.io.sql as psql
from sqlalchemy import create_engine
import re, timeit

%load_ext sql
#    format: %sql dialect+driver://username:password@host:port/database
engine = create_engine('postgresql://postgres:postgres@localhost/FapCov2103')
%sql postgresql://postgres:postgres@localhost/FapCov2103

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


Usando dados de D2

In [78]:
%%sql
SET Search_Path To D2; -- Usar dados dos hospitais com desfecho

 * postgresql://postgres:***@localhost/FapCov2103
Done.


[]

### Pivotamento

Técnica de __transformar colunas em tuplas__ é chamada: Pivotamento de relações (Retirado do Notebook 3.1)

In [79]:
%%sql
SELECT P.id_paciente, E.dt_coleta,
       MAX(E.de_hospital) AS Hospital,
       AVG(regexp_replace(E.de_resultado, ',', '.')::FLOAT) FILTER(WHERE E.de_analito ~*'ldl' AND E.de_analito!~*'vldl') AS LDL,      -- evitar contar vldl como ldl.
       AVG(regexp_replace(E.de_resultado, ',', '.')::FLOAT) FILTER(WHERE E.de_analito ~*'hdl') AS HDL, 
       AVG(regexp_replace(E.de_resultado, ',', '.')::FLOAT) FILTER(WHERE E.de_analito ~*'v.*coles')AS VLDL, 
       AVG(regexp_replace(E.de_resultado, ',', '.')::FLOAT) FILTER(WHERE E.de_analito ~*'n[aã]o.hdl')AS NaoHDL,                       -- aceita nao e não
       AVG(regexp_replace(E.de_resultado, ',', '.')::FLOAT) FILTER(WHERE E.de_analito IN('Colesterol total', 'COLESTEROL')) AS Total, -- alguns hospitais usam só colesterol para indicar total.
       MAX(E.cd_unidade) AS Unidade
    FROM D2.ExamLabs E JOIN D2.Pacientes P on E.id_paciente = P.ID_Paciente
    WHERE 
    -- P.CD_Municipio IN('GUARULHOS', 'OSASCO') AND
         E.de_exame ~*'coleste' AND 
         E.de_resultado!~'[^\d.,+-]'         -- pode haver texto ao invés de números em alguns registros.
    GROUP BY P.id_paciente, E.dt_coleta
    LIMIT 10;

 * postgresql://postgres:***@localhost/FapCov2103
10 rows affected.


[('00083836F6B96D4A', datetime.date(2020, 6, 22), 'BPSP', 132.0, 125.5, 19.0, 151.0, 251.0, 'mg/dL'),
 ('000CDFE696A34C62', datetime.date(2020, 9, 16), 'BPSP', 175.0, 142.0, 31.0, 206.0, 284.0, 'mg/dL'),
 ('000CDFE696A34C62', datetime.date(2021, 1, 21), 'BPSP', 159.0, 121.5, 23.0, 182.0, 243.0, 'mg/dL'),
 ('000F0BC139D2846DB86AA32B8F05B215', datetime.date(2020, 9, 24), 'HSL', 180.0, 123.5, 28.0, 208.0, 247.0, 'mg/dL'),
 ('000F0BC139D2846DB86AA32B8F05B215', datetime.date(2021, 2, 11), 'HSL', 171.0, 117.5, 26.0, 197.0, 235.0, 'mg/dL'),
 ('0024EA87F47A6563', datetime.date(2020, 7, 6), 'BPSP', 70.0, 76.5, 13.0, 83.0, 153.0, 'mg/dL'),
 ('002B919CC409B11DE52FB212379BE2CB', datetime.date(2020, 9, 23), 'HSL', 152.0, 123.0, 15.0, 167.0, 246.0, 'mg/dL'),
 ('003059D681358F66', datetime.date(2021, 1, 5), 'BPSP', 67.0, 45.0, 14.0, None, 126.0, 'mg/dL'),
 ('0034BAD8E9987AC7', datetime.date(2020, 8, 5), 'BPSP', 72.0, 54.0, 20.0, None, 146.0, 'mg/dL'),
 ('00371004112B005E', datetime.date(2020, 3, 18), 'BPSP', 176.0, 47.0, 17.0, None, 240.0, 'mg/dL')]

### Criando Visão Materializada de exam_labs para Colesterol

In [80]:
%%sql
DROP MATERIALIZED VIEW IF EXISTS colesterol CASCADE;
CREATE MATERIALIZED VIEW colesterol AS 
	SELECT P.id_paciente, E.dt_coleta,
       MAX(E.de_hospital) AS Hospital,
       AVG(regexp_replace(E.de_resultado, ',', '.')::FLOAT) FILTER(WHERE E.de_analito ~*'ldl' AND E.de_analito!~*'vldl') AS LDL,      -- evitar contar vldl como ldl.
       AVG(regexp_replace(E.de_resultado, ',', '.')::FLOAT) FILTER(WHERE E.de_analito ~*'hdl') AS HDL, 
       AVG(regexp_replace(E.de_resultado, ',', '.')::FLOAT) FILTER(WHERE E.de_analito ~*'v.*coles')AS VLDL, 
       AVG(regexp_replace(E.de_resultado, ',', '.')::FLOAT) FILTER(WHERE E.de_analito ~*'n[aã]o.hdl')AS NaoHDL,                       -- aceita nao e não
       AVG(regexp_replace(E.de_resultado, ',', '.')::FLOAT) FILTER(WHERE E.de_analito IN('Colesterol total', 'COLESTEROL')) AS Total, -- alguns hospitais usam só colesterol para indicar total.
       MAX(E.cd_unidade) AS Unidade
    FROM D2.ExamLabs E JOIN D2.Pacientes P ON E.id_paciente = P.ID_Paciente
    WHERE
-- 		P.CD_Municipio IN('GUARULHOS', 'OSASCO') AND
		E.de_exame ~*'coleste' AND 
        E.de_resultado!~'[^\d.,+-]'         -- pode haver texto ao invés de números em alguns registros.
    GROUP BY P.id_paciente, E.dt_coleta;

 * postgresql://postgres:***@localhost/FapCov2103
Done.
13876 rows affected.


[]

### Criando Coluna analito_ref

In [81]:
%%sql
DROP VIEW IF EXISTS colesterol_analito_ref;
CREATE VIEW colesterol_analito_ref AS
    (SELECT *, 
        ((case when ldl    < 130 then 1 else 0 end)    + 
         (case when hdl    > 40  then 1 else 0 end)*2  +
         (case when vldl   < 30  then 1 else 0 end)*4  + 
         (case when naohdl < 160 then 1 else 0 end)*8  +
         (case when total  < 190 then 1 else 0 end)*16)::bit(5) as analito_ref
        FROM colesterol
    );

SELECT * FROM colesterol_analito_ref LIMIT 20;

 * postgresql://postgres:***@localhost/FapCov2103
Done.
Done.
20 rows affected.


[('00083836F6B96D4A', datetime.date(2020, 6, 22), 'BPSP', 132.0, 125.5, 19.0, 151.0, 251.0, 'mg/dL', '01110'),
 ('000CDFE696A34C62', datetime.date(2020, 9, 16), 'BPSP', 175.0, 142.0, 31.0, 206.0, 284.0, 'mg/dL', '00010'),
 ('000CDFE696A34C62', datetime.date(2021, 1, 21), 'BPSP', 159.0, 121.5, 23.0, 182.0, 243.0, 'mg/dL', '00110'),
 ('000F0BC139D2846DB86AA32B8F05B215', datetime.date(2020, 9, 24), 'HSL', 180.0, 123.5, 28.0, 208.0, 247.0, 'mg/dL', '00110'),
 ('000F0BC139D2846DB86AA32B8F05B215', datetime.date(2021, 2, 11), 'HSL', 171.0, 117.5, 26.0, 197.0, 235.0, 'mg/dL', '00110'),
 ('0024EA87F47A6563', datetime.date(2020, 7, 6), 'BPSP', 70.0, 76.5, 13.0, 83.0, 153.0, 'mg/dL', '11111'),
 ('002B919CC409B11DE52FB212379BE2CB', datetime.date(2020, 9, 23), 'HSL', 152.0, 123.0, 15.0, 167.0, 246.0, 'mg/dL', '00110'),
 ('003059D681358F66', datetime.date(2021, 1, 5), 'BPSP', 67.0, 45.0, 14.0, None, 126.0, 'mg/dL', '10111'),
 ('0034BAD8E9987AC7', datetime.date(2020, 8, 5), 'BPSP', 72.0, 54.0, 20.0, None, 146.0, 'mg/dL', '10111'),
 ('00371004112B005E', datetime.date(2020, 3, 18), 'BPSP', 176.0, 47.0, 17.0, None, 240.0, 'mg/dL', '00110'),
 ('00371004112B005E', datetime.date(2020, 4, 1), 'BPSP', None, None, None, None, 217.0, 'mg/dL', '00000'),
 ('00371004112B005E', datetime.date(2020, 5, 3), 'BPSP', 148.0, 25.0, None, None, 202.0, 'mg/dL', '00000'),
 ('004688799FD293C3ABE0A07209FD8B75', datetime.date(2020, 5, 9), 'HSL', 95.0, 36.0, 27.0, None, 158.0, 'mg/dL', '10101'),
 ('004688799FD293C3ABE0A07209FD8B75', datetime.date(2020, 9, 8), 'HSL', 90.0, 36.0, 28.0, None, 154.0, 'mg/dL', '10101'),
 ('004688799FD293C3ABE0A07209FD8B75', datetime.date(2020, 10, 1), 'HSL', 174.0, 41.0, 43.0, None, 258.0, 'mg/dL', '00010'),
 ('004688799FD293C3ABE0A07209FD8B75', datetime.date(2020, 11, 23), 'HSL', 198.0, 67.0, 33.0, None, 298.0, 'mg/dL', '00010'),
 ('004688799FD293C3ABE0A07209FD8B75', datetime.date(2021, 2, 10), 'HSL', 106.0, 36.0, 28.0, None, 170.0, 'mg/dL', '10101'),
 ('004A3731E98C1598', datetime.date(2020, 7, 30), 'BPSP', 81.0, 68.0, 13.0, 94.0, 136.0, 'mg/dL', '11111'),
 ('00589831CDA58DE8', datetime.date(2020, 7, 15), 'BPSP', 105.0, 76.5, None, 125.0, 153.0, 'mg/dL', '11011'),
 ('005B118C512EE0B624AB7293F42B1D54', datetime.date(2020, 6, 26), 'HSL', 95.0, 53.0, 17.0, None, 165.0, 'mg/dL', '10111')]

### Contabilizando

In [82]:
%%sql
DROP VIEW IF EXISTS colesterol_analito_count;
CREATE VIEW colesterol_analito_count AS
    (SELECT analito_ref, COUNT(*) FROM colesterol_analito_ref
    GROUP BY analito_ref
    ORDER BY analito_ref
    ); 

SELECT * FROM colesterol_analito_count

 * postgresql://postgres:***@localhost/FapCov2103
Done.
Done.
27 rows affected.


[('00000', 338),
 ('00001', 148),
 ('00010', 1045),
 ('00011', 267),
 ('00100', 99),
 ('00101', 19),
 ('00110', 1088),
 ('00111', 312),
 ('01010', 28),
 ('01011', 217),
 ('01110', 319),
 ('01111', 973),
 ('10000', 254),
 ('10001', 757),
 ('10010', 58),
 ('10011', 456),
 ('10100', 20),
 ('10101', 901),
 ('10110', 11),
 ('10111', 1577),
 ('11001', 2),
 ('11010', 5),
 ('11011', 686),
 ('11100', 1),
 ('11101', 8),
 ('11110', 18),
 ('11111', 4269)]

### Solução

#### Obtendo valor de ajuste

Queremos 10% de 13876 valores. Existem 32 classes no nosso esquema.
Note que 13876/320 ~= 43.

In [89]:
%%sql
SELECT *, 
     (case when count < 43 then 43-count else 0 end) AS sobra,
     (case when count < 43 then 0     else 1 end) AS colunas_com_sobra
FROM colesterol_analito_count

 * postgresql://postgres:***@localhost/FapCov2103
27 rows affected.


[('00000', 338, 0, 1),
 ('00001', 148, 0, 1),
 ('00010', 1045, 0, 1),
 ('00011', 267, 0, 1),
 ('00100', 99, 0, 1),
 ('00101', 19, 24, 0),
 ('00110', 1088, 0, 1),
 ('00111', 312, 0, 1),
 ('01010', 28, 15, 0),
 ('01011', 217, 0, 1),
 ('01110', 319, 0, 1),
 ('01111', 973, 0, 1),
 ('10000', 254, 0, 1),
 ('10001', 757, 0, 1),
 ('10010', 58, 0, 1),
 ('10011', 456, 0, 1),
 ('10100', 20, 23, 0),
 ('10101', 901, 0, 1),
 ('10110', 11, 32, 0),
 ('10111', 1577, 0, 1),
 ('11001', 2, 41, 0),
 ('11010', 5, 38, 0),
 ('11011', 686, 0, 1),
 ('11100', 1, 42, 0),
 ('11101', 8, 35, 0),
 ('11110', 18, 25, 0),
 ('11111', 4269, 0, 1)]

In [84]:
%%sql
SELECT SUM(count) AS Total, SUM(sobra) AS count_sobra, SUM(classe_com_sobra) FROM
     (SELECT *, 
          (case when count < 43 then 43-count else 0 end) AS sobra,
          (case when count < 43 then 0     else 1 end) AS classe_com_sobra
     FROM colesterol_analito_count) AS query_sobra;

 * postgresql://postgres:***@localhost/FapCov2103
1 rows affected.


[(Decimal('13876'), Decimal('275'), 18)]

275/18 ~= 15
Portanto devem ser escolhidos 15 valores a mais (43+15 = 58 valores) nas classes com sobras.
Nas classes sem sobras, todos os valores são amostrados.

In [90]:
%%sql
DROP VIEW IF EXISTS colesterol_analito_ajuste;
CREATE VIEW colesterol_analito_ajuste AS
SELECT analito_ref,
     (case when colunas_com_sobra = 1 and count > 57 then 58 else count end)::float / count::float AS valor_ajuste
FROM (SELECT *, 
          (case when count < 43 then 0 else 1 end) AS colunas_com_sobra
     FROM colesterol_analito_count) as view_count;

SELECT * FROM colesterol_analito_ajuste;

 * postgresql://postgres:***@localhost/FapCov2103
Done.
Done.
27 rows affected.


[('00000', 0.17159763313609466),
 ('00001', 0.3918918918918919),
 ('00010', 0.05550239234449761),
 ('00011', 0.21722846441947566),
 ('00100', 0.5858585858585859),
 ('00101', 1.0),
 ('00110', 0.05330882352941176),
 ('00111', 0.1858974358974359),
 ('01010', 1.0),
 ('01011', 0.2672811059907834),
 ('01110', 0.18181818181818182),
 ('01111', 0.05960945529290853),
 ('10000', 0.2283464566929134),
 ('10001', 0.07661822985468957),
 ('10010', 1.0),
 ('10011', 0.12719298245614036),
 ('10100', 1.0),
 ('10101', 0.06437291897891231),
 ('10110', 1.0),
 ('10111', 0.03677869372225745),
 ('11001', 1.0),
 ('11010', 1.0),
 ('11011', 0.08454810495626822),
 ('11100', 1.0),
 ('11101', 1.0),
 ('11110', 1.0),
 ('11111', 0.013586319981260248)]

### Obtendo amostragem

In [97]:
%%sql
DROP TABLE IF EXISTS Exercicio6_Solucao;
CREATE TABLE Exercicio6_Solucao AS
SELECT * FROM colesterol_analito_ref WHERE False;
DO $$ DECLARE valor record;
BEGIN
    FOR valor IN SELECT analito_ref, valor_ajuste FROM colesterol_analito_ajuste
    LOOP
        INSERT INTO Exercicio6_Solucao 
        SELECT * FROM colesterol_analito_ref
        WHERE (analito_ref = valor.analito_ref AND Random() < valor.valor_ajuste);
    END LOOP;
END
$$;

SELECT * FROM Exercicio6_Solucao
ORDER BY Random()
LIMIT 20;

 * postgresql://postgres:***@localhost/FapCov2103
Done.
0 rows affected.
Done.
20 rows affected.


[('8A55DECB1056125A', datetime.date(2020, 10, 26), 'BPSP', 84.0, 51.0, None, None, 151.0, 'mg/dL', '10011'),
 ('9661B15E8EE807FC', datetime.date(2020, 6, 8), 'BPSP', 84.0, 85.5, None, 103.0, 171.0, 'mg/dL', '11011'),
 ('54160D652AE44081', datetime.date(2021, 1, 8), 'BPSP', None, None, None, None, 67.0, 'mg/dL', '10000'),
 ('47B5CFB8B45E155A', datetime.date(2020, 8, 5), 'BPSP', 129.0, 138.5, None, 154.0, 277.0, 'mg/dL', '01011'),
 ('8640F9695AA6E6B3D57E3F84715B7246', datetime.date(2021, 5, 17), 'HSL', 202.0, 30.0, 18.0, None, 250.0, 'mg/dL', '00100'),
 ('E4D7EA0C633D3213', datetime.date(2020, 5, 14), 'BPSP', 87.0, None, None, None, 144.0, 'mg/dL', '10001'),
 ('A8A30C8611F4B8DC', datetime.date(2020, 10, 6), 'BPSP', 55.0, 27.0, 19.0, None, 101.0, 'mg/dL', '10101'),
 ('D86E31A5C26FAE91', datetime.date(2020, 9, 21), 'BPSP', 211.0, 26.0, 42.0, None, 279.0, 'mg/dL', '00000'),
 ('C7B5F4F4EA9F6E8A25056B097827FB07', datetime.date(2021, 5, 1), 'HSL', 119.0, 46.0, 25.0, None, 190.0, 'mg/dL', '00111'),
 ('B0DF62E631843092', datetime.date(2020, 12, 3), 'BPSP', 82.0, 63.333333333333336, 18.0, 99.0, 144.0, 'mg/dL', '11111'),
 ('04B6CA07C252FEB97B52DE8E8C1AC973', datetime.date(2021, 3, 10), 'HSL', 123.0, 45.0, 24.0, None, 192.0, 'mg/dL', '00111'),
 ('5DBE504044E7B02CEAC2423305B4399B', datetime.date(2020, 9, 21), 'HSL', 91.0, 65.0, 15.0, None, 171.0, 'mg/dL', '10111'),
 ('595E058D5096D3B1', datetime.date(2020, 11, 20), 'BPSP', 85.0, 42.0, None, None, 151.0, 'mg/dL', '10011'),
 ('BAABC0DB40F84AF7', datetime.date(2021, 1, 16), 'BPSP', 23.0, 39.0, 25.0, 48.0, 78.0, 'mg/dL', '11101'),
 ('E30BBC5453C4B4F2', datetime.date(2021, 1, 28), 'BPSP', 117.0, 101.0, 22.0, 139.0, 202.0, 'mg/dL', '01111'),
 ('A26CF024252CD8D5', datetime.date(2020, 6, 9), 'BPSP', 107.0, 95.5, 23.0, 130.0, 191.0, 'mg/dL', '01111'),
 ('6621058AF933C523', datetime.date(2020, 10, 22), 'BPSP', 74.0, 69.0, None, 99.0, 138.0, 'mg/dL', '11011'),
 ('FBE33D6C2E5A5AC3917C17E174B6B988', datetime.date(2020, 11, 10), 'HSL', 172.0, 117.0, 33.0, 205.0, 234.0, 'mg/dL', '00010'),
 ('18BAE859DB0942927BC3D54DE6BC8270', datetime.date(2020, 8, 31), 'HSL', 116.0, 49.0, 32.0, None, 197.0, 'mg/dL', '00011'),
 ('AE8F2C60FE9DD245', datetime.date(2020, 5, 8), 'BPSP', 175.0, 122.5, 33.0, 208.0, 245.0, 'mg/dL', '00010')]

In [98]:
%%sql
SELECT analito_ref, COUNT(*) FROM Exercicio6_Solucao
GROUP BY analito_ref
ORDER BY analito_ref;

 * postgresql://postgres:***@localhost/FapCov2103
27 rows affected.


[('00000', 73),
 ('00001', 52),
 ('00010', 76),
 ('00011', 60),
 ('00100', 58),
 ('00101', 19),
 ('00110', 71),
 ('00111', 56),
 ('01010', 28),
 ('01011', 71),
 ('01110', 60),
 ('01111', 54),
 ('10000', 56),
 ('10001', 61),
 ('10010', 58),
 ('10011', 69),
 ('10100', 20),
 ('10101', 56),
 ('10110', 11),
 ('10111', 54),
 ('11001', 2),
 ('11010', 5),
 ('11011', 59),
 ('11100', 1),
 ('11101', 8),
 ('11110', 18),
 ('11111', 46)]

In [99]:
%%sql
SELECT COUNT(*) FROM Exercicio6_Solucao;

 * postgresql://postgres:***@localhost/FapCov2103
1 rows affected.


[(1202,)]